pkg/k8s/watchers/cilium_node.go

```diff
@@ -4,108 +4,85 @@
+
 						if n.IsLocal() {
 							return
 						}
 						k.nodeDiscoverManager.NodeUpdated(n)
-						k.K8sEventProcessed(metricCiliumNode, metricCreate, errs == nil)
-					}
-				},
-				UpdateFunc: func(oldObj, newObj interface{}) {
-					var valid, equal bool
-					defer func() { k.K8sEventReceived(metricCiliumNode, metricUpdate, valid, equal) }()
-					if oldCN := k8s.ObjToCiliumNode(oldObj); oldCN != nil {
-						if ciliumNode := k8s.ObjToCiliumNode(newObj); ciliumNode != nil {
-							valid = true
-							if oldCN.DeepEqual(ciliumNode) {
-								equal = true
-								return
-							}
-							n := nodeTypes.ParseCiliumNode(ciliumNode)
-							errs := k.CiliumNodeChain.OnUpdateCiliumNode(oldCN, ciliumNode, swgNodes)
-							if n.IsLocal() {
-								return
-							}
-							k.nodeDiscoverManager.NodeUpdated(n)
-							k.K8sEventProcessed(metricCiliumNode, metricUpdate, errs == nil)
-						}
-					}
-				},
-				DeleteFunc: func(obj interface{}) {
-					var valid, equal bool
-					defer func() { k.K8sEventReceived(metricCiliumNode, metricDelete, valid, equal) }()
-					ciliumNode := k8s.ObjToCiliumNode(obj)
-					if ciliumNode == nil {
-						return
-					}
-					valid = true
-					n := nodeTypes.ParseCiliumNode(ciliumNode)
-					errs := k.CiliumNodeChain.OnDeleteCiliumNode(ciliumNode, swgNodes)
-					if errs != nil {
-						valid = false
+						k.K8sEventProcessed(metricCiliumNode, metricUpdate, errs == nil)
 					}
-					k.nodeDiscoverManager.NodeDeleted(n)
-				},
+				}
 			},
-			k8s.ConvertToCiliumNode,
-		)
-		isConnected := make(chan struct{})
-		// once isConnected is closed, it will stop waiting on caches to be
-		// synchronized.
-		k.blockWaitGroupToSyncResources(isConnected, swgNodes, ciliumNodeInformer.HasSynced, k8sAPIGroupCiliumNodeV2)
-
-		once.Do(func() {
-			// Signalize that we have put node controller in the wait group
-			// to sync resources.
-			asyncControllers.Done()
-		})
-		k.k8sAPIGroups.AddAPI(k8sAPIGroupCiliumNodeV2)
-		go ciliumNodeInformer.Run(isConnected)
-
-		<-kvstore.Connected()
-		close(isConnected)
-
-		log.Info("Connected to key-value store, stopping CiliumNode watcher")
+			DeleteFunc: func(obj interface{}) {
+				var valid, equal bool
+				defer func() { k.K8sEventReceived(metricCiliumNode, metricDelete, valid, equal) }()
+				ciliumNode := k8s.ObjToCiliumNode(obj)
+				if ciliumNode == nil {
+					return
+				}
+				valid = true
+				n := nodeTypes.ParseCiliumNode(ciliumNode)
+				errs := k.CiliumNodeChain.OnDeleteCiliumNode(ciliumNode, swgNodes)
+				if errs != nil {
+					valid = false
+				}
+				k.nodeDiscoverManager.NodeDeleted(n)
+			},
+		},
+		k8s.ConvertToCiliumNode,
+	)
 
-		k.cancelWaitGroupToSyncResources(k8sAPIGroupCiliumNodeV2)
-		k.k8sAPIGroups.RemoveAPI(k8sAPIGroupCiliumNodeV2)
-		// Create a new node controller when we are disconnected with the
-		// kvstore
-		<-kvstore.Client().Disconnected()
+	k.ciliumNodeStoreMU.Lock()
```

Comment:
There's no need to protect `k.ciliumNodeStore` with a lock if we guarantee that `k.ciliumNodeStore` is set before we execute `func (k *K8sWatcher) updateCiliumNodeLabels(node *v1.Node)`

Reply:
I don't think there's any guarantee that `k.ciliumNodeInit()` will not run concurrently with the `sync-node-labels-with-ciliumnode` `Node` controller, so we should keep protecting this